File: armorforge/sources/tab_scene.ts

let tab_scene_line_counter: i32 = 0;
let tab_scene_new_meshes: string[] = null;
let _tab_scene_paint_object_length: i32 = 1;

function tab_scene_select_object(mo: mesh_object_t) {
	if (mo == null) {
		return;
	}

	context_raw.selected_object = mo.base;

	if (mo.base.ext_type != "mesh_object_t") {
		return;
	}

	context_raw.paint_object = mo;
	if (context_raw.merged_object != null) {
		context_raw.merged_object.base.visible = false;
	}
	context_select_paint_object(mo);
}

function tab_scene_import_mesh_done() {
	let count: i32 = project_paint_objects.length - _tab_scene_paint_object_length;
	_tab_scene_paint_object_length = project_paint_objects.length;

	for (let i: i32 = 0; i < count; ++i) {
		let mo: mesh_object_t = project_paint_objects[project_paint_objects.length - 1 - i];
		object_set_parent(mo.base, null);
		tab_scene_select_object(mo);
	}

	app_notify_on_next_frame(function() {
		util_mesh_merge();
	});
}

function tab_scene_draw_list(ui: ui_t, list_handle: ui_handle_t, current_object: object_t) {
	if (char_at(current_object.name, 0) == ".") {
		return; // Hidden
	}

	let b: bool = false;

	// Highlight every other line
	if (tab_scene_line_counter % 2 == 0) {
		g2_set_color(ui.ops.theme.SEPARATOR_COL);
		g2_fill_rect(0, ui._y, ui._window_w, ui_ELEMENT_H(ui));
		g2_set_color(0xffffffff);
	}

	// Highlight selected line
	if (current_object == context_raw.selected_object) {
		g2_set_color(0xff205d9c);
		g2_fill_rect(0, ui._y, ui._window_w, ui_ELEMENT_H(ui));
		g2_set_color(0xffffffff);
	}

	if (current_object.children.length > 0) {
		let row: f32[] = [1 / 13, 12 / 13];
		ui_row(row);
		let h: ui_handle_t = ui_nest(list_handle, tab_scene_line_counter);
		if (h.init) {
			h.selected = true;
		}
		b = ui_panel(h, "", true, false);
		ui_text(current_object.name);
	}
	else {
		ui._x += 18; // Sign offset

		// Draw line that shows parent relations
		g2_set_color(ui.ops.theme.BUTTON_COL);
		g2_draw_line(ui._x - 10, ui._y + ui_ELEMENT_H(ui) / 2, ui._x, ui._y + ui_ELEMENT_H(ui) / 2);
		g2_set_color(0xffffffff);

		ui_text(current_object.name);
		ui._x -= 18;
	}

	tab_scene_line_counter++;
	// Undo applied offset for row drawing caused by end_element()
	ui._y -= ui_ELEMENT_OFFSET(ui);

	if (ui.is_released) {
		tab_scene_select_object(current_object.ext);
	}

	if (ui.is_hovered && ui.input_released_r) {
		tab_scene_select_object(current_object.ext);

		ui_menu_draw(function (ui: ui_t) {
			if (ui_menu_button(ui, tr("Duplicate"))) {
				sim_duplicate();
			}
			if (ui_menu_button(ui, tr("Delete"))) {
				sim_delete();
			}
		});
	}

	if (b) {
		let current_y: i32 = ui._y;
		for (let i: i32 = 0; i < current_object.children.length; ++i) {
			let child: object_t = current_object.children[i];
			ui._x += 8;
			tab_scene_draw_list(ui, list_handle, child);
			ui._x -= 8;
		}

		// Draw line that shows parent relations
		g2_set_color(ui.ops.theme.BUTTON_COL);
		g2_draw_line(ui._x + 14, current_y, ui._x + 14, ui._y - ui_ELEMENT_H(ui) / 2);
		g2_set_color(0xffffffff);
	}
}

function tab_scene_new_object(mesh_name: string) {
	let blob: buffer_t = iron_load_blob(data_path() + "meshes/" + mesh_name);
	let raw: scene_t = armpack_decode(blob);
	util_mesh_ext_pack_uvs(raw.mesh_datas[0].vertex_arrays[2].values);
	let md: mesh_data_t = mesh_data_create(raw.mesh_datas[0]);
	md._.handle = md.name;
	let mo: mesh_object_t = scene_add_mesh_object(md, project_paint_objects[0].materials);
	mo.base.name = md.name;
	let o: obj_t = {};
	o._ = { _gc: raw };
	mo.base.raw = o;
	map_set(data_cached_meshes, md._.handle, md);
	array_push(project_paint_objects, mo);
	tab_scene_import_mesh_done();
	app_notify_on_next_frame(function(mo: mesh_object_t) {
		tab_scene_select_object(mo);
	}, mo);
}

function tab_scene_new_menu(ui: ui_t) {
	for (let i: i32 = 0; i < tab_scene_new_meshes.length; ++i) {
		let mesh_name: string = tab_scene_new_meshes[i];
		if (ui_menu_button(ui, mesh_name)) {
			tab_scene_new_object(mesh_name);
		}
	}
}

function tab_scene_draw(htab: ui_handle_t) {
	let ui: ui_t = ui_base_ui;
	if (ui_tab(htab, tr("Scene"))) {

		ui_begin_sticky();

		let row: f32[] = [1 / 4, 1 / 4];
		ui_row(row);

		if (ui_button("New")) {
			if (tab_scene_new_meshes == null) {
				tab_scene_new_meshes = file_read_directory(path_data() + path_sep + "meshes");
			}

			ui_menu_draw(tab_scene_new_menu);
		}

		if (ui_button("Import")) {
			project_import_mesh(false, tab_scene_import_mesh_done);
		}

		ui_end_sticky();

		{
			ui._y -= ui_ELEMENT_OFFSET(ui);

			tab_scene_line_counter = 0;

			let scene: object_t = _scene_root.children[0];

			for (let i: i32 = 0; i < scene.children.length; ++i) {
				let c: object_t = scene.children[i];
				tab_scene_draw_list(ui, ui_handle(__ID__), c);
			}

			// Select object with arrow keys
			if (ui.is_key_pressed && ui.key_code == key_code_t.DOWN) {
				let i: i32 = array_index_of(project_paint_objects, context_raw.selected_object.ext);
				if (i < project_paint_objects.length - 1) {
					tab_scene_select_object(project_paint_objects[i + 1]);
				}
			}
			if (ui.is_key_pressed && ui.key_code == key_code_t.UP) {
				let i: i32 = array_index_of(project_paint_objects, context_raw.selected_object.ext);
				if (i > 1) {
					tab_scene_select_object(project_paint_objects[i - 1]);
				}
			}
		}
	}
}

File: paint/sources/tab_scene.c

#include "global.h"

i32 tab_scene_line_counter = 0;

void tab_scene_select_object(mesh_object_t *mo) {
	if (mo == NULL) {
		return;
	}

	g_context->selected_object = mo->base;

	if (!string_equals(mo->base->ext_type, "mesh_object_t")) {
		return;
	}

	g_context->paint_object = mo;
	if (g_context->merged_object != NULL) {
		g_context->merged_object->base->visible = false;
	}
	context_select_paint_object(mo);
}

i32 tab_scene_sort_compare(void **pa, void **pb) {
	object_t *a = *(pa);
	object_t *b = *(pb);
	return strcmp(a->name, b->name);
}

void tab_scene_sort() {
	object_t *scene = _scene_root->children->buffer[0];
	array_sort(scene->children, &tab_scene_sort_compare);
}

void tab_scene_import_mesh_done_on_next_frame(void *_) {
	util_mesh_merge(NULL);
	tab_scene_select_object(g_context->selected_object->ext);
	tab_scene_sort();
}

void tab_scene_import_mesh_done() {
	i32 count                      = project_paint_objects->length - _tab_scene_paint_object_length;
	_tab_scene_paint_object_length = project_paint_objects->length;

	for (i32 i = 0; i < count; ++i) {
		mesh_object_t *mo = project_paint_objects->buffer[project_paint_objects->length - 1 - i];
		object_set_parent(mo->base, NULL);
		tab_scene_select_object(mo);
	}

	sys_notify_on_next_frame(&tab_scene_import_mesh_done_on_next_frame, NULL);
}

void tab_scene_draw_list_context_menu() {
	if (ui_menu_button(tr("Duplicate"), "", ICON_DUPLICATE)) {
		sim_duplicate();
	}
	if (ui_menu_button(tr("Delete"), "", ICON_DELETE)) {
		sim_delete();
	}
}

void tab_scene_draw_list(ui_handle_t *list_handle, object_t *current_object) {
	if (string_equals(char_at(current_object->name, 0), ".")) {
		return; // Hidden
	}

	bool b = false;

	// Highlight every other line
	if (tab_scene_line_counter % 2 == 0) {
		draw_set_color(ui->ops->theme->SEPARATOR_COL);
		draw_filled_rect(0, ui->_y, ui->_window_w, UI_ELEMENT_H());
		draw_set_color(0xffffffff);
	}

	// Highlight selected line
	if (current_object == g_context->selected_object) {
		draw_set_color(0xff205d9c);
		draw_filled_rect(0, ui->_y, ui->_window_w, UI_ELEMENT_H());
		draw_set_color(0xffffffff);
	}

	if (current_object->children->length > 0) {
		f32_array_t *row = f32_array_create_from_raw(
		    (f32[]){
		        1 / 13.0,
		        12 / 13.0,
		    },
		    2);
		ui_row(row);
		ui_handle_t *h = ui_nest(list_handle, tab_scene_line_counter);
		if (h->init) {
			h->b = true;
		}
		b = ui_panel(h, "", true, false, false);
		ui_text(current_object->name, UI_ALIGN_LEFT, 0x00000000);
	}
	else {
		ui->_x += 18; // Sign offset

		// Draw line that shows parent relations
		draw_set_color(ui->ops->theme->BUTTON_COL);
		draw_line(ui->_x - 10, ui->_y + UI_ELEMENT_H() / 2.0, ui->_x, ui->_y + UI_ELEMENT_H() / 2.0, 1.0);
		draw_set_color(0xffffffff);

		ui_text(current_object->name, UI_ALIGN_LEFT, 0x00000000);
		ui->_x -= 18;
	}

	tab_scene_line_counter++;
	// Undo applied offset for row drawing caused by end_element()
	ui->_y -= UI_ELEMENT_OFFSET();

	if (ui->is_released) {
		tab_scene_select_object(current_object->ext);
	}

	if (ui->is_hovered && ui->input_released_r) {
		tab_scene_select_object(current_object->ext);

		ui_menu_draw(&tab_scene_draw_list_context_menu, -1, -1);
	}

	if (b) {
		i32 current_y = ui->_y;
		for (i32 i = 0; i < current_object->children->length; ++i) {
			object_t *child = current_object->children->buffer[i];
			ui->_x += 8;
			tab_scene_draw_list(list_handle, child);
			ui->_x -= 8;
		}

		// Draw line that shows parent relations
		draw_set_color(ui->ops->theme->BUTTON_COL);
		draw_line(ui->_x + 14, current_y, ui->_x + 14, ui->_y - UI_ELEMENT_H() / 2.0, 1.0);
		draw_set_color(0xffffffff);
	}
}

void tab_scene_draw(ui_handle_t *htab) {
	if (ui_tab(htab, tr("Scene"), false, -1, false)) {

		tab_scene_line_counter = 0;

		object_t *scene = _scene_root->children->buffer[0];
		for (i32 i = 0; i < scene->children->length; ++i) {
			object_t *c = scene->children->buffer[i];
			tab_scene_draw_list(ui_handle(__ID__), c);
		}

		// Select object with arrow keys
		if (ui->is_key_pressed && ui->key_code == KEY_CODE_DOWN) {
			i32 i = array_index_of(project_paint_objects, g_context->selected_object->ext);
			if (i < project_paint_objects->length - 1) {
				tab_scene_select_object(project_paint_objects->buffer[i + 1]);
			}
		}
		if (ui->is_key_pressed && ui->key_code == KEY_CODE_UP) {
			i32 i = array_index_of(project_paint_objects, g_context->selected_object->ext);
			if (i > 1) {
				tab_scene_select_object(project_paint_objects->buffer[i - 1]);
			}
		}
	}
}
